### PR TITLE
clean up save states

### PR DESCRIPTION
**Pull request by MauricioUyaguari**
## Summary

- [x] fix bug with dataspace context being lost
- [x] clean up save states to avoid duplicate code

## How did you test this change?

- [ ] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

File: packages/legend-graph/src/graph-manager/action/query/Query.ts [hunk 1]
```diff
@@ -51,7 +51,6 @@ export class Query {
   lastUpdatedAt?: number | undefined;
   isCurrentUserQuery = false;
 }
-
```
Comment (akphi) [inline packages/legend-graph/src/graph-manager/action/query/Query.ts:54]: let's revert this change

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 2]
```diff
@@ -1003,9 +758,76 @@ export class ServiceQueryCreatorStore extends QueryEditorStore {
   }
 }
 
+export class ExistingQueryUpdateState {
+  readonly editorStore: ExistingQueryEditorStore;
+  queryRenamer = false;
+  updateQueryState = ActionState.create();
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:764]:
```suggestion
  readonly updateQueryState = ActionState.create();
```

File: packages/legend-application-query/src/components/QueryEditor.tsx [hunk 3]
```diff
@@ -324,49 +352,26 @@ const QueryEditorHeaderContent = observer(
...
+        return (
+          <QueryEditorExistingQueryHeader
+            queryBuilderState={queryBuilderState}
+            existingEditorStore={editorStore}
+          />
+        );
       }
+      return (
+        <div className="query-editor__header__content__main query-editor__header__content__title">
+          {editorStore.label}
```
Comment (akphi) [inline packages/legend-application-query/src/components/QueryEditor.tsx:370]: Nice idea, but I'm not sure the current labels are useful or not

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQuerySetupStore.ts [hunk 4]
```diff
@@ -155,8 +155,8 @@ export class DataSpaceQuerySetupStore extends QueryEditorStore {
     return true;
   }
 
-  override get isSaveActionDisabled(): boolean {
-    return true;
+  override get label(): string {
+    return 'DataSpace';
```
Comment (akphi) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQuerySetupStore.ts:159]: I'm not sure if this label is useful or not for the user, maybe something else, I'm not sure what, but this definitely won't help.

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts [hunk 5]
```diff
@@ -109,6 +109,10 @@ export class DataSpaceQueryCreatorStore extends QueryEditorStore {
     };
   }
 
+  override get label(): string {
+    return this.dataSpacePath;
```
Comment (akphi) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts:113]: This label won't help user neither, maybe something like `Data Space: ${dataSpace.title/name}` or something?

Comment (akphi) [reply packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts:113]: or something suggest that we're creating something new... ?

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 6]
```diff
@@ -611,32 +322,30 @@ export abstract class QueryEditorStore {
         },
       },
     );
+    this.queryCreatorState = new QueryCreatorState(this, undefined);
   }
 
-  get isViewProjectActionDisabled(): boolean {
-    return false;
-  }
+  abstract get label(): string;
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:328]: if it's not always useful/appropriate to show this, should we make it optional? e.g. the default implementation always return undefined, a plugin/extension can override if needs be

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 7]
```diff
@@ -646,6 +355,44 @@ export abstract class QueryEditorStore {
...
 
+  async buildQueryForStore(
+    query: Query,
+    rawLambda: RawLambda,
+    config: QueryPersistConfiguration | undefined,
+  ): Promise<Query> {
+    try {
+      assertNonNullable(
+        this._queryBuilderState,
+        'QueryBuilderState required to build query to edit',
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:366]:
```suggestion
        'Query builder state required to build query to edit',
```

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 8]
```diff
@@ -646,6 +355,44 @@ export abstract class QueryEditorStore {
...
+      );
+      assertNonNullable(
+        this._queryBuilderState.mapping,
+        'Query required mapping to update',
+      );
+      const runtimeValue = guaranteeType(
+        this._queryBuilderState.runtimeValue,
+        RuntimePointer,
+        'Query Runtime required to be of type RuntimePointer',
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:375]:
```suggestion
        'Query runtime must be of type runtime pointer',
```

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 9]
```diff
@@ -646,6 +355,44 @@ export abstract class QueryEditorStore {
     // do nothing
   }
 
+  async buildQueryForStore(
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:358]: sounds a bit confusing; `buildQueryForPersistence` or `assembleQuery`?

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 10]
```diff
@@ -853,6 +600,10 @@ export class MappingQueryCreatorStore extends QueryEditorStore {
     };
   }
 
+  override get label(): string {
+    return this.mappingPath;
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:604]: This will confuse people.

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 11]
```diff
@@ -611,32 +322,30 @@ export abstract class QueryEditorStore {
...
+      this._queryBuilderState,
+      'QueryBuilderState has not been set',
+    );
   }
 
   setExistingQueryName(val: string | undefined): void {
     this.existingQueryName = val;
   }
 
+  get isPerformingBlockingAction(): boolean {
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:345]: shouldn't we make this `mobx.computed` and the children `mobx.override`?

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 12]
```diff
@@ -1003,9 +758,76 @@ export class ServiceQueryCreatorStore extends QueryEditorStore {
...
+      this.editorStore.applicationStore.notificationService.notifyError(error);
+    } finally {
+      this.updateQueryState.reset();
+    }
+  }
+}
+
 export class ExistingQueryEditorStore extends QueryEditorStore {
   private queryId: string;
   private _query?: LightQuery | undefined;
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:828]: --> `_lightQuery` to be less confusing?

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 13]
```diff
@@ -527,28 +247,19 @@ export abstract class QueryEditorStore {
...
-  title: string | undefined;
+  _queryBuilderState?: QueryBuilderState | undefined;
+  queryCreatorState: QueryCreatorState;
   existingQueryName: string | undefined;
 
   constructor(
     applicationStore: LegendQueryApplicationStore,
     depotServerClient: DepotServerClient,
   ) {
     makeObservable(this, {
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:258]: should we add `_queryBuilderState: observable` and `queryBuilderState: computed`?

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 14]
```diff
@@ -611,32 +322,30 @@ export abstract class QueryEditorStore {
...
-    this.saveAsState = val;
-  }
-
-  setSaveState(val: QuerySaveState | undefined): void {
-    this.saveState = val;
-  }
-
-  setRenameState(val: QueryRenameState | undefined): void {
-    this.renameState = val;
+  get queryBuilderState(): QueryBuilderState {
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:334]: this change and the change to `_queryBuilderState` actually make it much harder to follow the code, I don't think `QueryEditorStore.queryBuilderState` gets used in too many places, can we just leave it like before? Thanks!

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts [hunk 15]
```diff
@@ -109,6 +109,10 @@ export class DataSpaceQueryCreatorStore extends QueryEditorStore {
     };
   }
 
+  override get label(): string | undefined {
+    return `Create Query with DataSpace ${this.dataSpacePath}`;
```
Comment (akphi) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts:113]: TODO: something to think about: in terms of styling, this is currently displayed the same way as name of a query, what should we do here?

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 16]
```diff
@@ -611,32 +327,23 @@ export abstract class QueryEditorStore {
         },
       },
     );
+    this.queryCreatorState = new QueryCreatorState(this, undefined);
   }
 
-  get isViewProjectActionDisabled(): boolean {
-    return false;
-  }
+  abstract get label(): string | undefined;
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:333]: maybe provide a default implementation to return `undefined`?

**Review by akphi — CHANGES_REQUESTED** [review]
you missed a changeset it seems

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts (hunk 15 above)
Comment (akphi) [reply packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts:113]: if you can think of a way to address this quickly, would be great to add it to this PR, else, it's okay, we can deal with it later

Comment (MauricioUyaguari) [reply packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQueryCreatorStore.ts:113]: yeah this was already an issue before the PR as we were messing up the styling if not an existing query. We can deal with it later

File: packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQuerySetupStore.ts [hunk 17]
```diff
@@ -155,8 +155,8 @@ export class DataSpaceQuerySetupStore extends QueryEditorStore {
     return true;
   }
 
-  override get isSaveActionDisabled(): boolean {
-    return true;
+  override get label(): string | undefined {
```
Comment (akphi) [inline packages/legend-extension-dsl-data-space/src/stores/query/DataSpaceQuerySetupStore.ts:158]: TODO: remove this

File: packages/legend-application-query/src/stores/QueryEditorStore.ts [hunk 18]
```diff
@@ -611,32 +327,25 @@ export abstract class QueryEditorStore {
         },
       },
     );
+    this.queryCreatorState = new QueryCreatorState(this, undefined);
   }
 
-  get isViewProjectActionDisabled(): boolean {
-    return false;
+  get label(): string | undefined {
```
Comment (akphi) [inline packages/legend-application-query/src/stores/QueryEditorStore.ts:333]: TODO: I think we can remove this altogether and in the flow of `EditExistingQueryState` we show the renamer

**Review by akphi — APPROVED** [review]
